doc/neps/nep-0035-array-creation-dispatch-with-array-function.rst

```diff
@@ -317,6 +329,81 @@ downsides pointed out above we have decided to discard any changes on the Python
+``like=`` is not propagated to the downstream library, nevertheless, it's still
+possible to access it. This requires some changes in the downstream library's
+``__array_function__`` definition, where the ``self`` attribute is in practice
+that passed via ``like=``. This is the case because we use ``like=`` as the
+dispatching array, unlike other compute functions covered by NEP-18 that usually
+dispatch on the first positional argument.
+
+An example of such use is to create a new Dask array while preserving its
+backend type:
+
+.. code:: python
+    # Returns dask.array<array, shape=(3,), dtype=int64, chunksize=(3,), chunktype=cupy.ndarray>
+    np.asarray([1, 2, 3], like=da.array(cp.array(())))
+
+    # Returns a cupy.ndarray
+    type(np.asarray([1, 2, 3], like=da.array(cp.array(()))).compute())
+
+Note how above the array is backed by ``chunktype=cupy.ndarray``, and the
+resulting array after computing it is also a ``cupy.ndarray``. If Dask does
+not use the ``like=`` argument via the ``self`` attribute from
+``__array_function__``, the example above would be backed by ``numpy.ndarray``
+instead:
+
+.. code:: python
+    # Returns dask.array<array, shape=(3,), dtype=int64, chunksize=(3,), chunktype=numpy.ndarray>
+    np.asarray([1, 2, 3], like=da.array(cp.array(())))
+
+    # Returns a numpy.ndarray
+    type(np.asarray([1, 2, 3], like=da.array(cp.array(()))).compute())
+
+Given the library would need to rely on ``self`` attribute from
+``__array_function__`` to dispatch the function with the correct reference
+array, we suggest one of two alternatives:
+
+1. Introduce a list of functions in the downstream library that do support the
+   ``like=`` argument and pass ``like=self`` when calling the function; or
+2. Inspect whether the function's signature and verify whether it includes the
+   ``like=`` argument. Note that this may incur in a higher performance penalty
+   and assumes introspection is possible, which may not be if the function is
+   a C function.
+
+To make things clearer, let's take a look at how suggestion 2 could be
+implemented in Dask. The current relevant part of ``__array_function__``
+definition in Dask is seen below:
+
+.. code:: python
+    def __array_function__(self, func, types, args, kwargs):
+        # Code not relevant for this example here
+
+        # Dispatch ``da_func`` (da.asarray, for example) with *args and **kwargs
+        da_func(*args, **kwargs)
+
+And this is how the updated code would look like:
+
+.. code:: python
+    def __array_function__(self, func, types, args, kwargs):
+        # Code not relevant for this example here
+
+        # Inspect ``da_func``'s  signature and store keyword-only arguments
+        import inspect
+        kwonlyargs = inspect.getfullargspec(da_func).kwonlyargs
+
+        # If ``like`` is contained in ``da_func``'s signature, add ``like=self``
+        # to the kwargs dictionary.
+        if 'like' in kwonlyargs:
+            kwargs['like'] = self
```

Comment:
Seeing the dance with `getfullargspec` (which obviously could be replaced with something else though), I actually wonder if the always pass solution (forcing downstream to drop it if they do not want to add `like=` to their own wrapper) isn't the better solution after all?  Maybe we really need some opinions how downstream authors feel about being forced to drop the `kwarg`, @hameerabbasi is one of them, @shoyer, or @mhvk may also have opinions?

The issue is whether `like=arrlike` is passed as a kwarg to `__array_function__` or (as currently the case) dropped and access to `arrlike` would have to be achieved by falling back to `self`.

Reply:
> Seeing the dance with `getfullargspec` (which obviously could be replaced with something else though), I actually wonder if the always pass solution (forcing downstream to drop it if they do not want to add `like=` to their own wrapper) isn't the better solution after all? Maybe we really need some opinions how downstream authors feel about being forced to drop the `kwarg`, @hameerabbasi is one of them, @shoyer, or @mhvk may also have opinions?

FWIW, I would rather not force downstream libraries to drop `like` manually. The solution here is, IMO, much more convenient for the majority of libraries (e.g., CuPy, that won't need `like=` explicit nor have to do any changes on their side), while still providing at least a couple of different alternatives to libraries that _may_ have a use for `like=` (e.g., Dask). Take my opinion with a grain of salt, but I'm here trying to chime in as a maintainer for Dask's `__array_function__`, and potentially to CuPy as well.

Reply:
@seberg - for astropy's `Quantity`, I think we could fairly easily deal with it either way, though having the `like` argument might be slightly more convenient with the current setup. But being an `ndarray` subclass, we're not really the target audience for the `like` argument anyway: in the end, we generally just wrap the numpy function call instead of calling something else entirely. I can definitely see that for dask/cupy it would be easier to just call their own routines, passing all the arguments (and since it is guaranteed that the argument of `like` is `self`, that should be OK; while arguably `__array_function__` could have been a class method, I think we also use that it is not). That said, code simplicity does count!

Reply:
Thanks, I was just curious what you think. I don't mind this, and we could make it more convenient in the future by providing the list of these function (e.g. as a special attribute on the public symbol).

Reply:
I don't mind having this list either, on the contrary, I think it would be useful. Would you prefer to have that still in 1.20 or should we leave it as is while this is experimental and, assuming we'll transition this feature to stable, try to address it in 1.21?

Reply:
I think we consider adding it even now if it is useful to you.  My gutfeeling would be making the example effectively: `if getattr(public_api, "__array_function_like_is_self__", False): ...` with some bike shedding on the name?

It doesn't really strike me as a difficult decision to be made. Although we could add a test in NumPy that such an attribute is set if `inspect` detects a like argument.

Reply:
Ah, I just realized that tagging something on to a C-defined function may be a bit too involved...

Reply:
Yeah, exactly. I think we can't `inspect` functions like `np.array`, but maybe there's some way I don't know of.